components/usage/pkg/controller/reconciler.go

```diff
@@ -71,11 +74,59 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
 	if len(invalidInstances) > 0 {
 		log.WithField("invalid_workspace_instances", invalidInstances).Errorf("Detected %d invalid instances. These will be skipped in the current run.", len(invalidInstances))
 	}
-
 	log.WithField("workspace_instances", instances).Debug("Successfully loaded workspace instances.")
+
+	workspaces, err := u.loadWorkspaces(ctx, instances)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load workspaces for workspace instances in time range: %w", err)
+	}
+	status.Workspaces = len(workspaces)
+
 	return status, nil
 }
 
+type workspaceWithInstances struct {
+	Workspace db.Workspace
+	Instances []db.WorkspaceInstance
+}
+
+func (u *UsageReconciler) loadWorkspaces(ctx context.Context, instances []db.WorkspaceInstance) ([]workspaceWithInstances, error) {
+	var workspaceIDs []string
+	for _, instance := range instances {
+		workspaceIDs = append(workspaceIDs, instance.WorkspaceID)
+	}
+
+	workspaces, err := db.ListWorkspacesByID(ctx, u.conn, workspaceIDs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find workspaces for provided workspace instances: %w", err)
+	}
+
+	// Map workspaces to corresponding instances
+	workspacesWithInstancesByID := map[string]workspaceWithInstances{}
+	for _, workspace := range workspaces {
+		workspacesWithInstancesByID[workspace.ID] = workspaceWithInstances{
+			Workspace: workspace,
+		}
+	}
+
+	// We need to also add the instances to corresponding records, a single workspace can have multiple instances
+	for _, instance := range instances {
+		item, ok := workspacesWithInstancesByID[instance.WorkspaceID]
+		if !ok {
+			return nil, errors.New("encountered instance without a corresponding workspace record")
+		}
+		item.Instances = append(item.Instances, instance)
+	}
+
+	// Flatten results into a list
+	var workspacesWithInstances []workspaceWithInstances
+	for _, w := range workspacesWithInstancesByID {
+		workspacesWithInstances = append(workspacesWithInstances, w)
+	}
+
+	return workspacesWithInstances, nil
+}
+
 func (u *UsageReconciler) loadWorkspaceInstances(ctx context.Context, from, to time.Time) ([]db.WorkspaceInstance, []invalidWorkspaceInstance, error) {
 	log.Infof("Gathering usage data from %s to %s", from, to)
 	instances, err := db.ListWorkspaceInstancesInRange(ctx, u.conn, from, to)
```

Comment:
Reading the code below, I see that we ignore currently running instances (no stoppedTime).

But what if an instance is kept running for 30 hours, when the user has only 2h of usage left?

Wouldn't it be better to set an "artificial" stoppedTime to "right now" on the provisional usage record, and thus ensure we also count hours for currently running workspaces?

Reply:
> Reading the code below, I see that we ignore currently running instances (no stoppedTime).

Where do you see the logic for this? We list all, including currently running instances, validate that the records are sensible (start is before end, if set, and there is a start time) and then split the set into valid and invalid.

Reply:
The relevant code:

https://github.com/gitpod-io/gitpod/blob/3c63e7eb6bb3570d4dc5debf1c17938d6a0a3591/components/usage/pkg/controller/reconciler.go#L113-L123

My point: Still running instances are valid. They should create a usage record going from `creationTime` to `getCurrentTime()`, thus giving a correct idea of how many seconds have been used so far (even if the instance hasn't been stopped yet).

Reply:
`if instance.StoppedTime.IsSet() {` will be false when the Stop time is not present (see comment on L113). So this skips instances which are currently running. We only mark it as invalid if it has a Stop time, and stop time is before the start time which is semantically invalid.

Reply:
And I've also now added a unit tests to prove this is the case - https://github.com/gitpod-io/gitpod/pull/10495/files#diff-a27489be974eefc74571351b4369b7f8f37c870194369a0d6423362f5b56a67bR32

Reply:
Gah, sorry, I misread the comment and didn't fully read the code snippet I shared 🤦‍♂️ sorry for the noise, and thanks for adding a test!

Reply:
That's okay. Code review is as much about learning how something works for the reviewer as it is to ensure we're doing the right thing.